### PR TITLE
Actions

### PR DESCRIPTION
**Pull request by elizabethrego**
Documentation below, can rewrite for wiki:

## How to make your notifications interactive

First, some background. <b>Actions</b> are little buttons the user can slide to see on a notification, and tap to have something done in the background or foreground, depending on how it's configured. On iOS you can add up to four notification actions. When notifications are shown in banner form, only two will be shown, and all four can be shown in the alert form. Of course, you can have less than four or two, or none. My implementation is such that the actions will be shown in the order you've supplied them, so that only the first two will show in banner form (by default, the behavior is a little bit different, but this was the best way I could make the order of the actions predictable). On Android, you can have up to three actions, and the banner vs. alert differences don't apply.

In iOS, notification actions are associated with a <b>category</b>. This category has those actions attached to it, and is part of the notification payload. Both actions and categories are identified by an identifier string (shocking, I know). Android does not associate actions with a category, or even with their own identifier. But to make the action behavior consistent across both platforms, <b>you will still need to use both action identifiers and a category</b> for Android notifications. This also uses less memory on Android. 

To add actions to a notification, you add them and a category when you schedule it. The actions will be an array (containing action objects) passed to the 'actions' property, and the category will be a string passed to the 'category' property. You can reuse actions, perhaps if you've defined them outside of your call to schedule a notification, and pick up particular ones from that definition in your call. However, after you've passed an action with a particular identifier, you can't modify it in later notifications. If two action objects have the same identifier, the first one passed will be used, and the next will turn into the first one in the native code. The same applies to categories. That will probably be the biggest 'gotcha' with my implementation, but I think it is perfectly reasonable behavior. So please be aware.

You can allow a user to send a text response through a notification (iOS only, would love to implement this for Android as well when it is available). All you need to do to allow for this is to add a 'behavior' property to your action object with the value 'textInput'. If it's not a reply-inline action, don't include this property, or provide any other value. Additionally, you can modify the text of the 'Send' button, by adding a 'textInputSendTitle' property, which the value of what you'd like the button to say. On Android, for these actions, I recommend handling them the way Gmail handles their 'Reply' notification action (

Each action must include six properties to work properly on both iOS and Android. There are also two, totally optional, properties to set up iOS inline reply actions. All of these properties are listed below.

<ul>
<li><b>identifier</b>: (iOS & Android) A string containing the identifier of your action (read above to learn more about how this will be used)</li>
<li><b>title</b>: (iOS & Android) A string that will be shown on the action button itself.</li>
<li><b>icon</b>: (Android) A string used to locate a resource asset to be shown on the action button itself. Don't include the file extension (but it should be .png). Whatever works for you for the smallIcon property on the notification itself will be what works here, so look into that for reference.
<li><b>activationMode</b>: (iOS) A string equal to either 'background' or 'foreground'. This property determines whether your app will be brought to the foreground (and opened, if it isn't already) when the action is tapped (for 'foreground', obviously).</li>
<li><b>destructive</b>: (iOS) A boolean. If true, the action button will be shown with a red background, to signifiy that destructive changes will be made if clicked. It doesn't actually change how anything functions.</li>
<li><b>authenticationRequired</b>: (iOS) A boolean. If true, the user will be required to enter their passcode before the action is actually performed.</li>
<li><b>behavior</b>: (iOS, optional) A string that should equal 'textInput' if you would like the reply-inline functionality described above.</li>
<li><b>textInputSendTitle</b>: (iOS, optional) A string that will appear on the 'Send' button for reply-inline actions (instead of 'Send'). This is optional if your reply specifies the 'textInput' behavior, and totally unnecessary if not. Note: You won't find this property under this name in Apple's documentation; it's set a different way internally.</li>
</ul>

### Sample

```javascript
var actions = [ {
        identifier: 'SIGN_IN',
        title: 'Yes',
        icon: 'res://ic_signin',
        activationMode: 'background',
        destructive: false,
        authenticationRequired: true
    },
    {
       identifier: 'MORE_SIGNIN_OPTIONS',
       title: 'More Options',
       icon: 'res://ic_moreoptions',
       activationMode: 'foreground',
       destructive: false,
       authenticationRequired: false
    },
    {
       identifier: 'PROVIDE_INPUT',
       title: 'Provide Input',
       icon: 'ic_input',
       activationMode: 'background',
       destructive: false,
       authenticationRequired: false,
       behavior: 'textInput',
       textInputSendTitle: 'Reply'
}];

cordova.plugins.notification.local.schedule({
        id: 1,
        title: "Production Jour fixe",
        text: "Duration 1h",
        at: monday_9_am,
        actions: [actions[0], actions[1]],
        category: 'SIGN_IN_TO_CLASS'
});
```

Of course, above, you can include all of the normal properties passed when scheduling a notification. Just add actions and category to get interactivity.

## Handling actions

When a user taps a notification action, you obviously want to do something. So, to know which action they've tapped, you just need to capture the 'action' event. This will give you the notification, state, and data. The action identifier (and optional reply-inline response info text) are in the data variable, an object with 'identifier' and (optional) 'responseInfoText' propertie(s). You should know all of the action identifiers (after all, you passed them), and use conditions to determine what you'll for for each actions.

### Sample

```javascript
cordova.plugins.notification.local.on('action', function (notification, state, data) {
    var replyMessage;

    if (data.identifier === 'PROVIDE_INPUT') {
        replyMessage = data.responseInfoText;
        alert(replyMessage);
    } else if (data.identifier === 'SIGN_IN') {
        alert('You have been signed in!');
    } else if (data.identifier === 'MORE_SIGNIN_OPTIONS') {
        alert('(Pretend there are more signin options here, please.)');
    } 
});
```

Obviously, you can do something more interesting than adding alerts.

In the case that you've shared actions amongst different notifications, but want to perform something else depending on what actual notification it was tapped from, just use the notification variable to access the payload you passed in and determine which it is. You'll need to parse this too.

## For Ionic users

If you use ngCordova, you can add this snippet to the ngCordova.plugins.localNotification module under the event listener:

```javascript
$window.cordova.plugins.notification.local.on('action', function (notification, state, data) {
  $timeout(function () {
    $rootScope.$broadcast('$cordovaLocalNotification:action', notification, state, data);
  });
});
```

Then in your app, you can capture the event like this:

```javascript
$rootScope.$on('$cordovaLocalNotification:action', function(notification, state, data) {
    // use conditionals to choose actions to take like above, blah blah
});
```

### CODE REVIEW

File: src/android/ClickActivity.java [hunk 1]
```diff
@@ -42,15 +42,27 @@
...
-        LocalNotification.fireEvent("click", notification);
 
-        super.onClick(notification);
+        if (!notification.getOptions().isOngoing()) {
+            String event = notification.isRepeating() ? "clear" : "cancel";
+
+            if (actionIdentifier != null) {
+                LocalNotification.fireEvent("action", notification, actionIdentifier);
+
+                if (event == "clear") {
```
Comment (katzer) [inline src/android/ClickActivity.java]: `event.equals("clear")`

File: src/android/ClickActivity.java [hunk 2]
```diff
@@ -42,15 +42,27 @@
...
+                    notification.clear();
```
Comment (katzer) [inline src/android/ClickActivity.java]: That should be placed back into the super class

File: src/android/ClickActivity.java [hunk 3]
```diff
@@ -42,15 +42,27 @@
...
+                } else {
+                    notification.cancel();
```
Comment (katzer) [inline src/android/ClickActivity.java]: That should be placed back into the super class

File: src/android/notification/Action.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        return allNotificationActionCategories.get(actionCategoryIdentifier);
+    }
+
+    /**
+     * Persists a new local notification action.
+     *
+     * @param action
+     *      New action to be added
+     */
+    public static void addNotificationAction(Action action) {
```
Comment (katzer) [inline src/android/notification/Action.java]: All of the methods & variables can be shortened.

```java
public static void addAction(Action action) { ... } // For example
```

File: src/android/notification/ClickActivity.java [hunk 5]
```diff
@@ -40,12 +40,6 @@
     @Override
     public void onClick(Notification notification) {
         launchApp();
-
-        if (notification.isRepeating()) {
```
Comment (katzer) [inline src/android/notification/ClickActivity.java]: Why?

File: src/android/notification/Options.java [hunk 6]
```diff
@@ -139,6 +146,42 @@ private void parseAssets() {
         }
     }
 
+    /*
+     * Parse actions.
+     */
+    private void parseActions() {
```
Comment (katzer) [inline src/android/notification/Options.java]: Method is to complex ;)

File: src/ios/APPLocalNotification.m [hunk 7]
```diff
@@ -735,4 +756,31 @@ - (void) fireEvent:(NSString*)event notification:(UILocalNotification*)notificat
...
+ - (void) handleNotificationAction:(NSNotification*)notification
+ {
+     NSString* identifier = [notification object];
+     
+     NSDictionary* userInfo = notification.userInfo;
+     UILocalNotification *localNotification = [userInfo objectForKey:@"localNotification"];
+     
+     NSDictionary* responseInfo = [userInfo objectForKey:@"responseInfo"];
+     
+     NSDictionary* dataDict = [NSDictionary dictionaryWithObjectsAndKeys:identifier, @"identifier",
```
Comment (katzer) [inline src/ios/APPLocalNotification.m]: max length per line is 80

File: src/ios/APPLocalNotification.m [hunk 8]
```diff
@@ -735,4 +756,31 @@ - (void) fireEvent:(NSString*)event notification:(UILocalNotification*)notificat
...
+                               [responseInfo objectForKey:@"UIUserNotificationActionResponseTypedTextKey"], @"responseInfoText", nil];
+     NSData* jsonData = [NSJSONSerialization dataWithJSONObject:dataDict options:0 error:nil];
```
Comment (katzer) [inline src/ios/APPLocalNotification.m]: max length per line is 80

File: src/ios/AppDelegate+APPLocalNotificationAction.m [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+ #import "AppDelegate+APPLocalNotificationAction.h"
+
+ @implementation AppDelegate (APPLocalNotificationAction)
+
+/**
+ * Handle notification actions for iOS < 9.
+ */
+ - (void) application:(UIApplication *)application handleActionWithIdentifier:(NSString *)identifier forLocalNotificiation:(UILocalNotification *)notification completionHandler:(void(^)())completionHandler;
```
Comment (katzer) [inline src/ios/AppDelegate+APPLocalNotificationAction.m]: Stay dry and delegate to the other one

File: src/ios/UIApplication+APPLocalNotification.m [hunk 10]
```diff
@@ -67,14 +70,96 @@ - (void) registerPermissionToScheduleLocalNotifications
...
 
         [[UIApplication sharedApplication]
          registerUserNotificationSettings:settings];
     }
 }
 
+/**
+ * Persist all actions and categories for notifications, adding new ones if necessary.
+ */
+- (NSSet*) processNotificationInteractions:(NSArray*)interactions
```
Comment (katzer) [inline src/ios/UIApplication+APPLocalNotification.m]: @elizabethrego I think you can you see the problem with that method :)

File: src/ios/UIApplication+APPLocalNotification.m [hunk 11]
```diff
@@ -26,6 +26,9 @@
 
 @implementation UIApplication (APPLocalNotification)
 
+NSMutableDictionary *allNotificationActions = nil;
```
Comment (katzer) [inline src/ios/UIApplication+APPLocalNotification.m]: Why is that necessary?

**Comment by katzer** [comment]
@elizabethrego As first I want to say thank you! I had that in mind since iOS 8+Android 5 but never found the time. :+1: :+1: 

Beside my comments I have a few more points.
#### Action event
I don't see the need for an additional event. Its a _click_ event and that already exist. Instead each callback could contain an event object with metadata.
```js
on("click", function(notification, payload, e){
    // e might contain infos about the app state, the clicked action, a.s.o.
});
```
#### Categories
1. Why is it necessary to store all actions and categories?
2. Whats the relation between actions+category and registerPermission?
3. Even some properties are iOS specific would you please find a more generic name. For example `tappable:Boolean ` instead of `activationMode:"background|"foreground""`

File: src/android/notification/ClickActivity.java [hunk 12]
```diff
@@ -40,12 +40,6 @@
...
-            notification.clear();
-        } else {
-            notification.cancel();
```
Comment (katzer) [inline src/android/notification/ClickActivity.java]: Why removed?

Comment (elizabethrego) [inline src/android/notification/ClickActivity.java]: Ugh, I wish I remembered. I think that was back when I was trying to do things a different way, and it  never got moved back. So, not good reason, just my sloppiness. Sorry about that! I've fixed the this locally.

**Comment by elizabethrego** [comment]
@katzer Thank you so much for taking the time to review all of this code. And I want to apologize for the issues you've found - I'm very rusty on native Android development and Java, and this is my first time working with iOS and Objective-C at all, so that should explain some of the strangeness.

I created the additional 'action' event because I didn't want to introduce any user-facing changes to those who weren't going to use actions. If you're fine with it, I don't see any reason not to modify the original 'click' event.

To clarify, I'll remove all use of the action event and pass a object with the click event. When an action button is clicked, the object will have 'actionIdentifier' and optional 'replyText' properties, both of which may be null. In the future, it may be useful for other things as well. If I'm understanding what you said, the click event would now be captured as:

```javascript
cordova.plugins.notification.local.on("click", function (notification, state, e) {
    alert(e.actionIdentifier + " was chosen"); // for example
}); 
```

Is that correct?

To answer your questions:

1. *Read answer to 2 first, then come back to this.* We register permissions for iOS each time a notification is scheduled. Because it is when we register permission that we set the categories, if we do not persistent them between different calls to register permission, each time a notification is scheduled the categories (and thus actions) would get overridden. 
  * I just re-confirmed that this morning - they definitely *do* need to be stored. If not stored, if I schedule two notifications (separately, not at the same time) and wait to click on them until one is scheduled, the actions for the first one scheduled disappear. 
  * For Android, they definitely *don't* need to be stored, but to make the experience consistent across both platforms they do. I think it would be very confusing to schedule a notification that shows up with different actions for each platform. 

2. For iOS, actions are attached to categories, and the category is included in the notification payload.  But these categories must FIRST be registered with the notification settings. 
  * From Apple documentation: image
  * Because we make the call to register settings in the registerPermissionToScheduleLocalNotifications method, this is where the categories need to be set, rather than setting actions on the notification payload itself as we do for Android.

3. I'm not sure what you mean. Do you want me to change the property names for the action objects that as passed to the schedule call in Javascript? If so, that's no problem. If I'm wrong please let me know!

I'm working on fixing all of the issues you mentioned already, and will make the changes described above pending your confirmation that I have the right idea.

File: src/ios/APPLocalNotification.m (hunk 8 above)
Comment (elizabethrego) [inline src/ios/APPLocalNotification.m]: Looks like the length is going over 80 in lots of places, should this be fixed in all of my obj-c code?

**Comment by elizabethrego** [comment]
Hi, just checking in to see if you've had any time to review my changes.

File: src/ios/AppDelegate+APPLocalNotificationAction.m [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ */
+ - (void)application:(UIApplication *)application handleActionWithIdentifier:(NSString *)identifier forLocalNotification:(UILocalNotification *)notification completionHandler:(void(^)())completionHandler  
+ {
+    [self application:[UIApplication sharedApplication]
+        handleActionWithIdentifier:identifier
+        forLocalNotification:notification
+        withResponseInfo:nil
+        completionHandler:(void (^)())completionHandler];
+    
+    completionHandler();
```
Comment (katzer) [inline src/ios/AppDelegate+APPLocalNotificationAction.m]: Looks like the handler will called twice (here and in again in `handleActionWithIdentifier`)

**Comment by katzer** [comment]
@elizabethrego Merry Christmas!

Would you please change the branch from master to 0.9-dev. Breaking changes like the callback signature are ok.

**Comment by mrqaidi** [comment]
nice feature  :+1:   . i wish   background mode plugin also have this feature

File: src/ios/AppDelegate+APPLocalNotificationAction.m (hunk 13 above)
Comment (elizabethrego) [inline src/ios/AppDelegate+APPLocalNotificationAction.m]: Ah yes, thanks for catching that. Fixed.

**Comment by elizabethrego** [comment]
@katzer Merry Christmas to you too, although belated! :)

I had to create a new pull request to change the branch, #799.

To clarify from your comments on Dec 10, do you still want the property names to be made more generic?

**Comment by katzer** [comment]
> do you still want the property names to be made more generic?

Yes if you can find good naming (short names)

**Comment by katzer** [comment]
Closing it in favor of #799
